### ISSUE TITLE
Refactor MSTUMP

### ISSUE THREAD

**Issue by mexxexx** [TDAmeritrade/stumpy#135]
While working on issue #117 I realized that the MSTUMP module has some opportunities for improving code readability and consistency, as well as significantly improving performance. I want to discuss these points here:

#### _multi_mass 
Throughout the code we use the mass algorithm in two different ways: In `core.mass` it is used to calculate the distance profile between a query `Q` and a time series `T`. Then we have `stamp.mass` which first calculates the distance profile `D` between `Q` and a time series `T`, then applies an exclusion zone and in the end returns the minimum value of `D` as well as its position. This is then used to calculate a *single* value of the matrix profile.

So far so good. But now we also have `mstump._multi_mass`. Here we not only calculate a single value of the Matrix profile, but we return a whole array `P`. `P` is initialized with `np.inf` and later we find the lines
 
```
for i in range(d):
        D_prime = D_prime + D[i]
        D_prime_prime = D_prime / (i + 1)
        # Element-wise Min
        col_mask = P[i] > D_prime_prime #######
        P[i, col_mask] = D_prime_prime[col_mask]
        I[i, col_mask] = trivial_idx
```
So the line I marked with ####### will always return an array that is True everywhere (exept where `D_prime_prime=np.inf` but this case does not change anything), so effectively we are returning the d-dimensional distance profile between `Q` and `T` here. 
But at the same time we apply an exclusion zone
```
zone_start = max(0, trivial_idx - excl_zone)
zone_stop = min(k, trivial_idx + excl_zone)
D[:, zone_start : zone_stop + 1] = np.inf
```
which gives us something in between `core.mass` and `stamp.mass`. For better readability and extensibility I would suggest dropping the application of an exclusion zone and the calculation of the indices in the last line, so that `mstump._multi_mass` returns the actual d-dimensional distance profile and so that it is in line with its one dimensional brother in `core`. (Of course, then some changes have to be made, in particular in `_calculate_first_profile_index`)

#### sorting

The most time consuming part of `mstump` (that I saw using cProfile) is the sorting in lines 378-381
```
# Column-wise sort
for col in range(k):
    # row_idx[:, col] = np.argsort(D[:, col])
    # D[:, col] = D[row_idx[:, col], col]
    D[:, col] = np.sort(D[:, col])
```
as we iterate over the whole length of the profile. It is much much faster to use `np.sort(D, axis=0)`. Sadly numba does not support this yet, so I would suggest moving this to a non jitted function and use stock numpy. I saw a speed increase of about 5-10x in my real world application.

#### mstump

Another part that I find very confusing is the following (in `_mstump`)
```
for i in range(d):
    D_prime = D_prime + D[i]
    D_prime_prime = D_prime / (i + 1)
    # Element-wise Min
    for col in range(k): ########
        if P[i, col] > D_prime_prime[col]:
            P[i, col] = D_prime_prime[col]
            I[i, col] = idx
``` 
I see that this follows the original paper, but what has to be noted is that every time we update the whole matrix profile (line #######). This makes sense for an anytime algorithm how it was originally proposed, but mstump is not an anytime algorithm and cannot be stopped at will. So my suggestion is to remove this anytime capability. We calculate the d-dimensional distance profile of `T[idx:idx+m]` in lines 1-3 of the snippet. So the final multidimensional matrix profile value at position `idx` and dimension `i` will be `min(D_prime_prime)` in the i-th iteration,  and there is no need to update the profile values at any other position. This saves again a loop over the whole profile and will greatly increase performance. I also think it is much more readable this way, since it follows the implementations of `stomp` and `stump`.

I think it makes sense to implement `naive_mstump` using the approach from the paper to have a test that the results are as expected.

What do you think?

**Comment by seanlaw** [comment]
Regarding:

> _multi_mass

I am 100% on-board with you. Can you create a separate issue for this?

> sorting

Can you try: 

```
# Column-wise sort
for col in prange(k):  # <- Notice that it is "prange" and not "range"
    # row_idx[:, col] = np.argsort(D[:, col])
    # D[:, col] = D[row_idx[:, col], col]
    D[:, col] = np.sort(D[:, col])
```

This will do parallel sorting and might use all cores to sort each column independently. I'd like to see a test case for this as I find it really surprising that you are getting a 5-10x speedup. 

>  mstump

> So my suggestion is to remove this anytime capability.

Can you show me, with code, what you are proposing? I'm certainly open to it but I'd like to understand first.

**Comment by mexxexx** [comment]
> Can you show me, with code, what you are proposing? I'm certainly open to it but I'd like to understand first.

I'm thinkin something along these lines: 
```
def _mstump(T, m, range_stop, excl_zone, M_T, Σ_T, QT, QT_first, μ_Q, σ_Q, k, range_start=1):
    QT_odd = QT.copy()
    QT_even = QT.copy()
    d = T.shape[0]

    P = np.empty((d, range_stop - range_start), dtype=float)
    I = np.empty((d, range_stop - range_start), dtype=int)
    D = np.empty((d, k), dtype=float)
    D_prime = np.empty(k, dtype=float)

    for i in range(range_start, range_stop):
        _calculate_distance_profiles(
            T, m, D, M_T, Σ_T, μ_Q, σ_Q, QT_even, QT_odd, QT_first, i, d, k
        ) # distance profile for each dimension
        zone_start = max(0, idx - excl_zone)
        zone_stop = min(k, idx + excl_zone)
        D[:, zone_start : zone_stop + 1] = np.inf

        D = np.sort(D, axis=0) # or parallel sorting, this we'll have to benchmark

        D_prime[:] = 0.0
        for i in range(d):
            D_prime = D_prime + D[i]

            min_index = np.argmin(D_prime)
            I[i] = min_index
            P[i] = D_prime[min_index]
            if P[i] == np.inf:
                I[i] = -1

    for i in range(d):
        P[i] /= i + 1

    return P, I
```

**Comment by seanlaw** [comment]
This is pretty interesting. Having not gone through this with a fine-tooth comb, my only concern is with the memory consumption for large multi-dimensional inputs and performance. If we can confirm that for medium to large arrays (of size 5 rows x 100 columns, 5 rows x 1,000 columns, 5 rows x 10,000 columns, 5 rows x 100,000 columns):

1) This uses the same/less memory than the original
2) This is faster or equal in terms of computational time than the original
3) Tests still pass (without alteration)

If so, then we should be good. I like that we get to use `core._calculate_squared_distance_profile` but it is technically not parallelized. Above, I think we should really do:

`for i in prange(range_start, range_stop):` 

And this ensures that it will be embarrassingly parallelized but these are all things we should test as it relates to performance. According to the performance plot, a 100,000 length time series will take around 2 hours on a 2 core machine and so I expect it to take around 10 hours for 5 dimensions.

**Comment by mexxexx** [comment]
Actually we would use the same code we have right now already for `calculate_distance_profiles`. An approach like in stomp where you update the QT product on the go (and in parallel). I just moved this to a different function for brevity.

I would suggest to slowly try improving and evaluate every major step along the way for memory/performance questions.

**Comment by seanlaw** [comment]
Cool! Thanks for thinking through this so thoroughly.

**Comment by mexxexx** [comment]
Also one more question: We have `P.shape=(dimensions, profile_length)` and in mstump we return `P.T`. The documentation says

> P (ndarray) – The multi-dimensional matrix profile. Each row of the array corresponds to each matrix profile for a given dimension (i.e., the first row is the 1-D matrix profile and the second row is the 2-D matrix profile).

So shouldn't we return the untransposed version `P` instead? I think that in a 2d array you get the i-th row with `P[i, :]` 

Thanks for some clarification

**Comment by seanlaw** [comment]
@mexxexx Good question. The documentation is outdated and needs to get changed. In #104 I changed the output of `mstump`/`mstumped` because the format was inconsistent with `stump`/`stumped`.

So, it should now read:

```
P : ndarray
    The multi-dimensional matrix profile. Each column of the array corresponds to each matrix profile for a given dimension (i.e., the first column is the 1-D matrix profile and the second column is the 2-D matrix profile).

Similarly, the docstring for `I` should be:

> I : ndarray
      The multi-dimensional matrix profile index where each column of the array corresponds to each matrix profile index for a given dimension.

IIRC, the reason for this change (the output) is that I had added some support for pandas data frames (at least as an input type) and while the conversion of a 1-D NumPy output array to a `pd.Series` doesn't care if it is a row or column vector, the conversion of a 2-D NumPy output array to a `pd.DataFrame` does care about the orientation of the input array. At the end of the day, it made more sense for a Pandas array to have more rows than columns and so I was forced to return `P.T`. 

Does that make sense? Sorry for forgetting to update the docstring as I was likely too focused on getting `gpu_stump` running fast. You may be the first one to take a serious look at `mstump`/`mstumped` so I appreciate the fresh eyes. Would you mind updating the docs for `mstump` and `mstumped`?

**Comment by mexxexx** [comment]
Thank you. Is there anything to do apart from editing the comment in the code?

**Comment by seanlaw** [comment]
No, that should be it. The unit test should already account for the transpose (and hence why the tests are passing successfully). Thank you!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by seanlaw** [comment]
@mexxexx Should this be closed as well now?

**Comment by mexxexx** [comment]
I would like to keep this open as I think there is a bit more refactoring to do, and I would like to reference this issue.

**Comment by seanlaw** [comment]
No problem! Feel free to re-open any issues that I’ve accidentally closed. For some reason, your PRs aren’t automatically closing the issues that it references so I’ve been closing them manually. I wonder if the Git comment format is incorrect and Github is not parsing it correctly.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mexxexx** [comment]
I did some profiling regarding the sorting aspect (jupyter notebook in firefox, Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 4 threads):

```
def sort_numpy(T, d, k):
    return np.sort(T, axis=0)

@njit(parallel=True, fastmath=True)
def sort_numba(T, d, k):
    for i in prange(k):
        # T[:,i] = np.sort(T[:, i]) ## no assignment as we don't want to actually sort T
        np.sort(T[:, i])

T = np.random.uniform(-100, 100, (d, k))
```

I precompiled `sort_numba`. and got the following results:

`d=5`, `k =10000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
295 µs ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 100 sort_numba(T, d, k)
6.69 ms ± 30.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`d=5`, `k =100000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
3.08 ms ± 60.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 100 sort_numba(T, d, k)
66.2 ms ± 82.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`d=50`, `k =10000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
11.2 ms ± 81.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 1000 sort_numba(T, d, k)
6.58 ms ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So sorting with numpy is orders of magnitude faster for a small number of dimensions and long time serieses, the case that we usually have.

I will look into how the two methods affect performance when actually used in `_mstump`.

**Comment by seanlaw** [comment]
IIRC, neither NumPy nor Numba performs sorting in parallel (this is a hard problem). So, for a given `i`, Numba is still using a single core to sort that vector. Is this correct?

Separately, Is Numba actually leveraging multiple cores to sort each `i` independently from the other? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mexxexx** [comment]
> So, for a given i, Numba is still using a single core to sort that vector. Is this correct?

That's right I think.

> Separately, Is Numba actually leveraging multiple cores to sort each i independently from the other?

Yes, I saw that all cores were active when I did the profiling.

I tried `mstump` using the two different sorting techniques and I got the following results (`m=10`):
|   | `n=10000, d=5` | `n=50000, d=5` | `n=10000, d=20` | CPU usage |
| ------------- | ------------- | ------------- | ------------- | ------------- | 
| sort_numba | 69s  | 80s | 1741s |~390% |
| sort_numpy  | 9s | 49s | 268s |~210-220% |

I did a single run with every configuration, precompiled mstump and tried to leave my computer running without doing anything else. 

So while the numpy version does not parallelize well, it is still faster by a large margin. We should check if somehow the numpy sorting algorithm can be leveraged to numba, for example in parallel chunks.

**Comment by seanlaw** [comment]
That's a huge difference. According to the Numba documentation (and some conversations on their Gitter channel), it shouldn't be that slow. I've posted an issue on the Numba Github issue tracker.

**Comment by mexxexx** [comment]
I did some research and we can parallelize the function by our selfs. I managed to make a working `sort_numpy_parallelized` function by following exactly this approach: https://stackoverflow.com/a/39856310 It only uses the (standard) `multiprocessing`library.

It was using all cores that I specified and speed up the sorting of a `(5, 1000000)` array from around 80ms to 30ms using four cores.

**Comment by seanlaw** [comment]
> I did some research and we can parallelize the function by our selfs. I managed to make a working sort_numpy_parallelized function by following exactly this approach: https://stackoverflow.com/a/39856310 It only uses the (standard) multiprocessing library.

> It was using all cores that I specified and speed up the sorting of a (5, 1000000) array from around 80ms to 30ms using four cores.

Cool. We might want to slow down on this a little just in case the Numba folks have a nice solution for sorting out of the box. It's great that we can leverage multiprocessing (and not add a new dependency) but this has its drawbacks on Windows machines too.

**Comment by mexxexx** [comment]
Yes of course! I still wanted to collect the results here, for future reference.

Below is a working example of a parallelized sorting. For the size of 20x1e8 I did not notice a speed difference between four and eight cores, even though all of them were used (1.2s per call vs 3.9s for single core).

```
import multiprocessing
import ctypes
import numpy as np

import time

shared_array = None
def sort_np(s):
    #shared_array[:,s].sort(axis=0) # Use this if you want sorting to actually happend
    np.sort(shared_array[:, s], axis=0)

def main():
    d = 20
    n = 10000000
    
    shared_array_base = multiprocessing.Array(ctypes.c_double, d*n)
    
    global shared_array
    shared_array = np.frombuffer(shared_array_base.get_obj())
    shared_array = shared_array.reshape(d, n)
    
    num_cores = 4
    pool = multiprocessing.Pool(num_cores)
    
    step = 1 + n // num_cores
    slices = [slice(step * i, step * (i+1)) for i in range(num_cores)]
    
    shared_array[:] = np.random.uniform(-100, 100, (d, n))
        
    num_reps = 10
    
    start = time.time()
    #for i in range(num_reps):
    #    np.sort(shared_array, axis=0)
    end = time.time()
    print("Sequential per rep: {}".format((end-start) / num_reps))
    
    start = time.time()
    for i in range(num_reps):
        pool.map(sort_np, slices)
    end = time.time()
    print("Parallel per rep: {}".format((end-start) / num_reps))
    
    
if __name__ == '__main__':
    main()
```

**Comment by mexxexx** [comment]
In the meantime, what do you think about moving the numba sorting to a separate function. This way users can monkey-patch it if desired, and we can fix the other issues regarding MSTUMP (nan/inf handling and finishing refactoring) and do the sorting in a separate issue.

**Comment by seanlaw** [comment]
@mexxexx I would keep the sorting in its current function and create a different issue that specifically addresses improving sorting.
